### ISSUE TITLE
OpenSSL 1.1.0 support in stable branch

### ISSUE THREAD

**Issue by thmo** [randombit/botan#753]
Is building against OpenSSL 1.1.0 supported in the stable branch?

I am seeing compile errors:
```
src/engine/openssl/ossl_md.cpp:43:18: error: field 'md' has incomplete type 'EVP_MD_CTX {aka evp_md_ctx_st}'
       EVP_MD_CTX md;
                  ^~
In file included from /usr/include/openssl/evp.h:14:0,
                 from src/engine/openssl/ossl_md.cpp:9:
/usr/include/openssl/ossl_typ.h:92:16: note: forward declaration of 'EVP_MD_CTX {aka struct evp_md_ctx_st}'
 typedef struct evp_md_ctx_st EVP_MD_CTX;
                ^~~~~~~~~~~~~
src/engine/openssl/ossl_md.cpp: In destructor 'virtual Botan::{anonymous}::EVP_HashFunction::~EVP_HashFunction()':
src/engine/openssl/ossl_md.cpp:98:26: error: 'EVP_MD_CTX_cleanup' was not declared in this scope
    EVP_MD_CTX_cleanup(&md);
                          ^
src/engine/openssl/ossl_bc.cpp:43:30: error: field 'encrypt' has incomplete type 'EVP_CIPHER_CTX {aka evp_cipher_ctx_st}'
       mutable EVP_CIPHER_CTX encrypt, decrypt;
                              ^~~~~~~
In file included from /usr/include/openssl/evp.h:14:0,
                 from src/engine/openssl/ossl_bc.cpp:9:
/usr/include/openssl/ossl_typ.h:90:16: note: forward declaration of 'EVP_CIPHER_CTX {aka struct evp_cipher_ctx_st}'
 typedef struct evp_cipher_ctx_st EVP_CIPHER_CTX;
                ^~~~~~~~~~~~~~~~~
src/engine/openssl/ossl_bc.cpp:43:39: error: field 'decrypt' has incomplete type 'EVP_CIPHER_CTX {aka evp_cipher_ctx_st}'
       mutable EVP_CIPHER_CTX encrypt, decrypt;
                                       ^~~~~~~
In file included from /usr/include/openssl/evp.h:14:0,
                 from src/engine/openssl/ossl_bc.cpp:9:
/usr/include/openssl/ossl_typ.h:90:16: note: forward declaration of 'EVP_CIPHER_CTX {aka struct evp_cipher_ctx_st}'
 typedef struct evp_cipher_ctx_st EVP_CIPHER_CTX;
                ^~~~~~~~~~~~~~~~~
```

Botan 1.10.14, OpenSSL 1.1.0c

**Comment by randombit** [comment]
Actually AFAIK even Botan 1.11 has not been tested with OpenSSL 1.1.0 yet. Looks like OpenSSL making the structures opaque has broken some of our assumptions.

**Comment by randombit** [comment]
Checked my system, even Arch is still using OpenSSL 1.0.2 so I suspect none of our CI is testing against OpenSSL 1.1.0

**Comment by cordney** [comment]
Indeed, OpenSSL 1.1.0 has made all structs opaque, see https://wiki.openssl.org/index.php/1.1_API_Changes.  Supporting < 1.1.0 and 1.1.0 at the same time will probably require lots of ``#ifdef`` and/or wrapper functions like the OpenSSL wiki page suggests, but I didn't look into this deeper.

Regarding Botan 1.10, any chance we add OpenSSL 1.1.0 support ever? Since it's a bugfix only release line I expect we will only add it to the 2.x line (to be released in January).

**Comment by randombit** [comment]
Unless the patch ends up being very simple to backport I don't see 1.10.x supporting the new OpenSSL API. The 1.10 branch was released 5+ years ago and basically EOL, so anyone expecting it to work with a library whose API has recently incompatibly changed, breaking many/most? applications using it, does not seem realistic to me. It would be good to fix this for 2.x however.

**Comment by bkircher** [comment]
FYI, building 2.1.0 fails for me with OpenSSL 1.1.0f

> It would be good to fix this for 2.x however.

:+1:

**Comment by securitykernel** [comment]
Support for OpenSSL 1.1.0 was added just recently and will land in 2.2.0, see https://github.com/randombit/botan/blob/master/news.rst#version-220-not-yet-released.

I think this issue in general should be closed. I don't see any patch coming for Botan 1.10.

**Comment by bkircher** [comment]
> Support for OpenSSL 1.1.0 was added just recently and will land in 2.2.0

Great! Any idea when this will be, roughly?

**Comment by randombit** [comment]
> Great! Any idea when this will be, roughly?

A new release is already well overdue at this point. So probably within the next week or so.

**Comment by randombit** [comment]
Closing this - 2.2 is released and supports OpenSSL 1.1. Botan 1.10 does not/will not, and will now (in future release) `#error` out if it detects being built against OpenSSL 1.1.
